### ISSUE TITLE
DeepSeek-OCR training hits NaN gradients (query encoder / SAM)

### ISSUE THREAD

**Issue by bkpcoding** [bkpcoding/torchtitan#2]
## Summary
Single-GPU DeepSeek-OCR debug runs in the Singularity container complete 2 steps but produce NaN gradients. NaNs appear during backward, not forward. Disabling images removes NaNs; freezing SAM shifts NaNs to the query encoder.

## Environment
- Singularity: `singularity/4.1.5`
- CUDA: `12.6.3`
- Container: `torchtitan/models/deepseek_ocr/singularity/deepseek_ocr.sif`
- GPU: NVIDIA L40S

## Repro (latest)
```bash
module load singularity/4.1.5 cuda/12.6.3
HF_HOME=/tmp/hf_cache singularity exec --nv \
  --bind /gpfs/accounts/bucherb_owned_root/bucherb_owned1/bpatil/torchtitan:/opt/torchtitan \
  ./deepseek_ocr.sif \
  /bin/bash -lc 'TORCHTITAN_FREEZE_SAM=1 TORCHTITAN_DISABLE_SAM_POS_EMBED_GRAD=1 \
  LOCAL_RANK=0 RANK=0 WORLD_SIZE=1 MASTER_ADDR=127.0.0.1 MASTER_PORT=29500 \
  python /opt/torchtitan/torchtitan/train.py \
    --job.custom-config-module torchtitan.custom_data_config \
    --job.config-file /opt/torchtitan/torchtitan/models/deepseek_ocr/train_configs/debug.toml \
    --model.flavor=debugmodel_full_vision \
    --data.img_size=1024 \
    --training.steps=2 \
    --training.local_batch_size=1 \
    --training.mixed_precision_param=float32 \
    --training.mixed_precision_reduce=float32 \
    --activation_checkpoint.mode=none'
```

## Observations
- Forward pass is finite (pred + loss_sum are finite).
- First non-finite gradients occur in:
  - With images + no freeze: `sam_encoder.pos_embed` and `sam_encoder.patch_embed.proj.weight`.
  - With `TORCHTITAN_FREEZE_SAM=1`: `query_encoder.layers.0.self_attn.q_proj.weight`.
- With images disabled (`TORCHTITAN_DISABLE_OCR_IMAGES=1`), training is stable (loss/grad_norm finite).

## Debugging stats (first step)
- `labels[min,max]=(-100,121)` and `ignore_ratio ~0.91`
- `images[min,max,mean]=(-2.1179, 2.64, 0.474)`
- `pred[min,max] ~ (-4.7, 4.7)`
- `loss_sum ~ 1415` (finite)

## Files to inspect
- `torchtitan/models/deepseek_ocr/model/sam_encoder.py`
- `torchtitan/models/deepseek_ocr/model/query_encoder.py`
- `torchtitan/models/deepseek_ocr/model/model.py` (scatter, freezing)
- `torchtitan/models/deepseek_ocr/datasets/ocr_datasets.py` (collator + SynthDOG parsing)
- `torchtitan/train.py` (batch/pred/loss/grad logging)
- `torchtitan/distributed/utils.py` (disable AMP when float32)

## Local debug changes (for reference)
- Parse SynthDOG `ground_truth` JSON string if needed.
- OCR collator now returns `(input_dict, labels)` and shifts labels.
- Added logging: batch stats, pred stats, loss stats, first non-finite grad parameter.
- Added env toggles:
  - `TORCHTITAN_DISABLE_OCR_IMAGES=1`
  - `TORCHTITAN_FREEZE_SAM=1`
  - `TORCHTITAN_DISABLE_SAM_POS_EMBED_GRAD=1`

## Open questions
- Is query encoder attention mask or scaling causing gradient instability?
- Should SAM/query encoder use different init or LR?
- Are there expected constraints on image normalization for SAM?

**Comment by bkpcoding** [comment]
Update from local Singularity debugging (2026-02-14):\n\nI reproduced #2 using  and applied a stabilization pass.\n\nChanges made:\n- : switched mixed-attention mask to bool keep-mask and replaced masked SDPA path with a numerically guarded attention computation (float32 logits + stabilized softmax + nan_to_num).\n- : added input RMSNorm before the query transformer stack.\n- : reduced query encoder depth/MLP size for debug stability while keeping SAM-compatible hidden dims.\n- : added explicit  for container path correctness.\n\nObserved after patch:\n- : no non-finite gradient warning; 2-step run completes with finite loss.\n- without freezing SAM: NaN gradients are gone, but  is still  on step 1/2 (residual magnitude instability).\n\nNext I’m continuing on the  path (likely norm overflow / very large finite grads) and will post a follow-up with a targeted fix + verification logs.

**Comment by bkpcoding** [comment]
Follow-up update (2026-02-14):

I continued on the `grad_norm: inf` path and added a targeted fix in `torchtitan/distributed/utils.py`:

- `clip_grad_norm_` now falls back to a float64 norm computation when `get_total_norm` returns non-finite but gradients are finite.
- same float64 fallback was added for EP/non-EP branches in `_clip_grad_norm_with_ep`.

Why:
- We were seeing `grad_norm = inf` even when per-parameter gradients were finite.
- This is consistent with norm accumulation overflow in float32 for very large finite gradients.

Validation (Singularity, non-freeze SAM):
- command path: `torchtitan/models/deepseek_ocr/singularity/commands.md` (+ explicit `--model.hf_assets_path`)
- step 1: `grad_norm: 4905795280942678409216.0000` (finite)
- step 2: `grad_norm: 61.0305` (finite)
- run completes 2 steps, no NaN/non-finite gradient warning line.

Current status:
- NaN gradient issue from this ticket is resolved in my local repro.
- `grad_norm` no longer shows `inf`, but step-1 magnitude can still be extremely large; this is now a stability/tuning issue rather than numeric-overflow-to-inf.
